modules/network_synchronizer/networked_controller.cpp

```diff
@@ -0,0 +1,1112 @@
+	PlayerInputsReference pir;
+	pir.set_inputs_buffer(bit_array);
+
+	while (ofs < data_len) {
+
+		ERR_FAIL_COND_MSG(ofs + 1 > data_len, "The arrived packet size doesn't meet the expected size.");
+		// First byte is used for the duplication count.
+		const uint8_t duplication = p_data[ofs];
+		ofs += 1;
+
+		// Validate input
+		pir.inputs_buffer.seek(ofs * 8);
+		const int input_size_in_bits = node->call("count_input_size", &pir);
+		// Pad to 8 bits.
+		const int input_size =
+				Math::ceil((static_cast<float>(input_size_in_bits)) / 8.0);
+		ERR_FAIL_COND_MSG(ofs + input_size > data_len, "The arrived packet size doesn't meet the expected size.");
+
+		// The input is valid, populate the buffer.
+		for (int sub = 0; sub <= duplication; sub += 1) {
+
+			const uint64_t input_id = first_input_id + inserted_input_count;
+			inserted_input_count += 1;
+
+			if (current_input_buffer_id != UINT64_MAX && current_input_buffer_id >= input_id)
+				continue;
+
+			FrameSnapshotSkinny rfs;
+			rfs.id = input_id;
+
+			const bool found = std::binary_search(
+					snapshots.begin(),
+					snapshots.end(),
+					rfs,
+					is_remote_frame_A_older);
+
+			if (!found) {
+				rfs.inputs_buffer.get_bytes_mut().resize(input_size);
+				copymem(
+						rfs.inputs_buffer.get_bytes_mut().ptrw(),
+						p_data.ptr() + ofs,
+						input_size);
+
+				snapshots.push_back(rfs);
+
+				// Sort the new inserted snapshot.
+				std::sort(snapshots.begin(), snapshots.end(), is_remote_frame_A_older);
+			}
+		}
+
+		// We can now advance the offset.
+		ofs += input_size;
+	}
+
+	ERR_FAIL_COND_MSG(ofs != data_len, "At the end was detected that the arrived packet has an unexpected size.");
+}
+
+int ServerController::notify_input_checked(uint64_t p_input_id) {
+	ERR_PRINT("The method `notify_input_checked` must not be called on server. Be sure why it happened.");
+	return 0;
+}
+
+uint64_t ServerController::get_stored_input_id(int p_i) const {
+	ERR_PRINT("The method `get_input_id` must not be called on server. Be sure why it happened.");
+	return UINT64_MAX;
+}
+
+bool ServerController::process_instant(int p_i, real_t p_delta) {
+	ERR_PRINT("The method `process_instant` must not be called on server. Be sure why it happened.");
+	return false;
+}
+
+uint64_t ServerController::get_current_input_id() const {
+	return current_input_buffer_id;
+}
+
+int ServerController::get_inputs_count() const {
+	return snapshots.size();
+}
+
+bool ServerController::fetch_next_input() {
+	bool is_new_input = true;
+	bool is_packet_missing = false;
+
+	if (unlikely(current_input_buffer_id == UINT64_MAX)) {
+		// As initial packet, anything is good.
+		if (snapshots.empty() == false) {
+			node->set_inputs_buffer(snapshots.front().inputs_buffer);
+			current_input_buffer_id = snapshots.front().id;
+			snapshots.pop_front();
+		} else {
+			is_new_input = false;
+		}
+		// Never notify a packet is missing untill the first one is not yet
+		// arrived.
```

Comment:
```suggestion
		// Don't notify about missed packets until at least one packet has arrived.
```

Not sure if I got the meaning of the original sentence correctly, was a little hard to parse for me :)

Reply:
You understood it perfectly. Done